### PR TITLE
Add the `on_before_optimizer_step` hook

### LINKED ISSUE #7924: `on_after_backward` should always run after backward
## 🐛 Bug

If `accumulate_grad_batches` is enabled, we don't call `on_after_backward` until we step the optimizers

https://github.com/PyTorchLightning/pytorch-lightning/blob/d209b689796719d1ab4fcc8e1c26b8b57cd348c4/pytorch_lightning/trainer/training_loop.py#L757-L763

This means `on_after_backward` is acting like `on_before_optimizer_step`.

So we should add that and always run `on_after_backward` after backward.

### PR DESCRIPTION
**Pull request by ddrevicky**
## What does this PR do?

Adds `on_before_optimizer_step` hook which behaves as `on_after_backward` did, i.e., it is called only once before optimizer step and after all backward passes and after `on_after_backward`

co-author: @carmocca

Continues #8328 
Fixes #7924 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Comment by ddrevicky** [comment]
Hi @carmocca, this is still a draft but I wanted to get your feedback. 

### Question 1

I tried putting `on_before_optimizer_step` before this line https://github.com/PyTorchLightning/pytorch-lightning/blob/cdcc483e9b7a79de3e5a7ac9c1e9dfd12ab77f4f/pytorch_lightning/loops/training_batch_loop.py#L187 

so it looks like this:
```
if self.trainer.lightning_module.automatic_optimization:
    self.trainer.call_hook('on_before_optimizer_step', batch_idx, optimizer, opt_idx)       # Problem: called first
    self.optimizer_step(optimizer, opt_idx, batch_idx, closure)          # on_after_backward called after from within the closure
```
while calling `self.on_after_backward(batch_idx, result.loss)` right after this line https://github.com/PyTorchLightning/pytorch-lightning/blob/cdcc483e9b7a79de3e5a7ac9c1e9dfd12ab77f4f/pytorch_lightning/loops/training_batch_loop.py#L571

and putting it outside the `if not self.should_accumulate():` condition. If that's what you meant then there is a problem there that `on_before_optimizer_step` is called first and then `on_after_backward` is called from within the closure which is not what we want here if I understand correctly. I therefore put `on_before_optimizer_step` into the closure exactly where `on_after_backward` is currently (which makes sense to me since the whole point of the issue that it currently behaves like `on_before_optimizer_step`). Is that fine with you (seemed to work fine when I tried it with `accum_grad_batches`)?

### Question 2
Could you clear up for me what you meant by this quote (from the original issue)?

> Regardless, maybe we shouldn't call it in that case as it won't get called when using manual optimization without AMP.
> In those cases, it should be replaced with before_optimizer_step.

I would like to know whether I should modify the precision plugins in any way with regards to `on_before_optimizer_step` and `on_after_backward` and how exactly because I am not clear on how those work.

### TODO:
I will have to update the hook tests most likely since they're failing and also do a once over to check that the new hook is everywhere where it should be.

**Comment by carmocca** [comment]
> this is still a draft but I wanted to get your feedback.

Hey @ddrevicky! Sorry for my silence.

I've been making progress on improving the tests we have for the hook order and noticed several inconsistencies in how we call the backward and optimizer step hooks. Will be addressing those in https://github.com/PyTorchLightning/pytorch-lightning/pull/8071.

So I recommend you don't work on this PR until then, as the necessary changes for this will be greatly simplified and the tests for it too.

If you have any further questions hit me up in Slack! I can ping you when #8071 is merged to finish this, or update this PR myself then.

Thanks!

**Comment by ddrevicky** [comment]
No problem @carmocca. If it's fine by you, please finish this PR once you're done with #8071, it will be simpler than keeping me in the loop. Thanks :slightly_smiling_face:.

**Review by carmocca — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 1]
```diff
@@ -104,13 +104,14 @@ def post_backward(
...
         self,
-        pl_module: 'pl.LightningModule',
+        model: 'pl.LightningModule',
         optimizer: Optimizer,
         optimizer_idx: int,
         lambda_closure: Callable,
         **kwargs: Any,
     ) -> bool:
         """Hook to do something before each optimizer step."""
+        model.trainer.call_hook("on_before_optimizer_step", optimizer, optimizer_idx)
```
Comment (tchaton) [inline pytorch_lightning/plugins/precision/precision_plugin.py:114]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**